In [1]:
import requests
import pandas as pd
import json

## Make an API request that returns the months where "fire" was searched in 2016. Which month had the most searches?

In [30]:
endpoint = 'https://data.nashville.gov/resource/fuaa-r5cm.json?year=2016&query_text=fire'

response = requests.get(endpoint) 

In [31]:
response

<Response [200]>

In [32]:
response.text

'[{"month_name":"January","year":"2016","query_count":"19","query_text":"fire"}\n,{"month_name":"February","year":"2016","query_count":"35","query_text":"fire"}\n,{"month_name":"March","year":"2016","query_count":"32","query_text":"fire"}\n,{"month_name":"April","year":"2016","query_count":"26","query_text":"fire"}\n,{"month_name":"May","year":"2016","query_count":"24","query_text":"fire"}\n,{"month_name":"June","year":"2016","query_count":"31","query_text":"fire"}\n,{"month_name":"July","year":"2016","query_count":"24","query_text":"fire"}\n,{"month_name":"August","year":"2016","query_count":"47","query_text":"fire"}\n,{"month_name":"September","year":"2016","query_count":"36","query_text":"fire"}\n,{"month_name":"October","year":"2016","query_count":"38","query_text":"fire"}\n,{"month_name":"November","year":"2016","query_count":"32","query_text":"fire"}\n,{"month_name":"December","year":"2016","query_count":"38","query_text":"fire"}]\n'

In [33]:
searches = response.json()
searches

[{'month_name': 'January',
  'year': '2016',
  'query_count': '19',
  'query_text': 'fire'},
 {'month_name': 'February',
  'year': '2016',
  'query_count': '35',
  'query_text': 'fire'},
 {'month_name': 'March',
  'year': '2016',
  'query_count': '32',
  'query_text': 'fire'},
 {'month_name': 'April',
  'year': '2016',
  'query_count': '26',
  'query_text': 'fire'},
 {'month_name': 'May',
  'year': '2016',
  'query_count': '24',
  'query_text': 'fire'},
 {'month_name': 'June',
  'year': '2016',
  'query_count': '31',
  'query_text': 'fire'},
 {'month_name': 'July',
  'year': '2016',
  'query_count': '24',
  'query_text': 'fire'},
 {'month_name': 'August',
  'year': '2016',
  'query_count': '47',
  'query_text': 'fire'},
 {'month_name': 'September',
  'year': '2016',
  'query_count': '36',
  'query_text': 'fire'},
 {'month_name': 'October',
  'year': '2016',
  'query_count': '38',
  'query_text': 'fire'},
 {'month_name': 'November',
  'year': '2016',
  'query_count': '32',
  'query_text

### At first it would only pull the 2014 data? i needed to get all of the years so I found where to filter data on the Socrata website.

ANSWER 1: August 2016 with 47 searches for fire

## 2. Make an API request that returns all the times a query was run more than 100 times in a month. How many times did this occur?

endpoint = 'https://data.nashville.gov/resource/fuaa-r5cm.json'

params = {
    'min': 100,
    'max': 600
}

response = requests.get(endpoint, params = params)

response.text

In [58]:
endpoint = 'https://data.nashville.gov/resource/fuaa-r5cm.json?$where=query_count>100&$limit=44000'

response = requests.get(endpoint) 

In [59]:
response.text

'[{"month_name":"March","year":"2014","query_count":"101","query_text":"permits"}\n,{"month_name":"January","year":"2015","query_count":"101","query_text":"criminal court clerk"}\n,{"month_name":"September","year":"2015","query_count":"101","query_text":"codes"}\n,{"month_name":"March","year":"2016","query_count":"101","query_text":"police"}\n,{"month_name":"March","year":"2016","query_count":"101","query_text":"civil service"}\n,{"month_name":"November","year":"2016","query_count":"101","query_text":"jobs"}\n,{"month_name":"November","year":"2017","query_count":"101","query_text":"metro holidays"}\n,{"month_name":"November","year":"2017","query_count":"101","query_text":"longevity pay"}\n,{"month_name":"January","year":"2018","query_count":"101","query_text":"West Nashville Heights Church of Christ"}\n,{"month_name":"January","year":"2018","query_count":"101","query_text":"property maps"}\n,{"month_name":"March","year":"2018","query_count":"101","query_text":"public works"}\n,{"month_

In [60]:
searches = response.json()
searches

[{'month_name': 'March',
  'year': '2014',
  'query_count': '101',
  'query_text': 'permits'},
 {'month_name': 'January',
  'year': '2015',
  'query_count': '101',
  'query_text': 'criminal court clerk'},
 {'month_name': 'September',
  'year': '2015',
  'query_count': '101',
  'query_text': 'codes'},
 {'month_name': 'March',
  'year': '2016',
  'query_count': '101',
  'query_text': 'police'},
 {'month_name': 'March',
  'year': '2016',
  'query_count': '101',
  'query_text': 'civil service'},
 {'month_name': 'November',
  'year': '2016',
  'query_count': '101',
  'query_text': 'jobs'},
 {'month_name': 'November',
  'year': '2017',
  'query_count': '101',
  'query_text': 'metro holidays'},
 {'month_name': 'November',
  'year': '2017',
  'query_count': '101',
  'query_text': 'longevity pay'},
 {'month_name': 'January',
  'year': '2018',
  'query_count': '101',
  'query_text': 'West Nashville Heights Church of Christ'},
 {'month_name': 'January',
  'year': '2018',
  'query_count': '101',
 

In [61]:
len(searches)

1262

ANSWER 2: 1262

## 3. Make another API request that returns all the times "codes" was searched more than 100 times in a month. How many times did this occur?

In [64]:
endpoint = 'https://data.nashville.gov/resource/fuaa-r5cm.json?$where=query_count>100&$limit=44000&query_text=codes'

response = requests.get(endpoint) 

In [65]:
response

<Response [200]>

In [66]:
response.text
searches = response.json()
searches

[{'month_name': 'September',
  'year': '2015',
  'query_count': '101',
  'query_text': 'codes'},
 {'month_name': 'April',
  'month': '4',
  'year': '2021',
  'query_count': '102',
  'query_text': 'codes'},
 {'month_name': 'August',
  'year': '2016',
  'query_count': '104',
  'query_text': 'codes'},
 {'month_name': 'November',
  'month': '11',
  'year': '2020',
  'query_count': '104',
  'query_text': 'codes'},
 {'month_name': 'December',
  'year': '2017',
  'query_count': '106',
  'query_text': 'codes'},
 {'month_name': 'May',
  'year': '2016',
  'query_count': '107',
  'query_text': 'codes'},
 {'month_name': 'April',
  'year': '2016',
  'query_count': '111',
  'query_text': 'codes'},
 {'month_name': 'July',
  'month': '7',
  'year': '2021',
  'query_count': '114',
  'query_text': 'codes'},
 {'month_name': 'January',
  'year': '2017',
  'query_count': '115',
  'query_text': 'codes'},
 {'month_name': 'May',
  'month': '5',
  'year': '2021',
  'query_count': '115',
  'query_text': 'codes'

In [68]:
len(searches)

56

ANSWER 3: 56

## Make an API request that returns the entire Top 500 Monthly Searches dataset. Make a chart that shows the number of times "maps" was searched in a month across the entire time frame.

In [ ]:
#response_API = requests.get('https://data.nashville.gov/resource/fuaa-r5cm.json?$select=count(*)')
#print(response_API.status_code)

In [90]:
#https://data.cityofchicago.org/resource/tt4n-kn4t.json?$where=annual_salary between '40000' and '60000'

data = response_API.text
data

json.loads(data)

In [91]:
#full500 = pd.read_json('https://data.nashville.gov/resource/fuaa-r5cm.json')
#full500.tail()

In [109]:
endpoint = 'https://data.nashville.gov/resource/fuaa-r5cm.json?query_text=maps'

response_maps = requests.get(endpoint) 

In [110]:
response_maps

<Response [200]>

In [111]:
response_maps.text

'[{"month_name":"January","year":"2014","query_count":"97","query_text":"maps"}\n,{"month_name":"February","year":"2014","query_count":"140","query_text":"maps"}\n,{"month_name":"March","year":"2014","query_count":"144","query_text":"maps"}\n,{"month_name":"April","year":"2014","query_count":"152","query_text":"maps"}\n,{"month_name":"May","year":"2014","query_count":"119","query_text":"maps"}\n,{"month_name":"June","year":"2014","query_count":"121","query_text":"maps"}\n,{"month_name":"July","year":"2014","query_count":"121","query_text":"maps"}\n,{"month_name":"August","year":"2014","query_count":"118","query_text":"maps"}\n,{"month_name":"September","year":"2014","query_count":"115","query_text":"maps"}\n,{"month_name":"October","year":"2014","query_count":"100","query_text":"maps"}\n,{"month_name":"November","year":"2014","query_count":"82","query_text":"maps"}\n,{"month_name":"December","year":"2014","query_count":"97","query_text":"maps"}\n,{"month_name":"January","year":"2015","

In [112]:
searches_maps = response_maps.json()
searches_maps

[{'month_name': 'January',
  'year': '2014',
  'query_count': '97',
  'query_text': 'maps'},
 {'month_name': 'February',
  'year': '2014',
  'query_count': '140',
  'query_text': 'maps'},
 {'month_name': 'March',
  'year': '2014',
  'query_count': '144',
  'query_text': 'maps'},
 {'month_name': 'April',
  'year': '2014',
  'query_count': '152',
  'query_text': 'maps'},
 {'month_name': 'May',
  'year': '2014',
  'query_count': '119',
  'query_text': 'maps'},
 {'month_name': 'June',
  'year': '2014',
  'query_count': '121',
  'query_text': 'maps'},
 {'month_name': 'July',
  'year': '2014',
  'query_count': '121',
  'query_text': 'maps'},
 {'month_name': 'August',
  'year': '2014',
  'query_count': '118',
  'query_text': 'maps'},
 {'month_name': 'September',
  'year': '2014',
  'query_count': '115',
  'query_text': 'maps'},
 {'month_name': 'October',
  'year': '2014',
  'query_count': '100',
  'query_text': 'maps'},
 {'month_name': 'November',
  'year': '2014',
  'query_count': '82',
  'q

In [136]:
maps_df = pd.DataFrame(searches_maps)
maps_df

month_name  year query_count query_text month
0     January  2014          97       maps   NaN
1    February  2014         140       maps   NaN
2       March  2014         144       maps   NaN
3       April  2014         152       maps   NaN
4         May  2014         119       maps   NaN
..        ...   ...         ...        ...   ...
77   November  2020          86       maps    11
78   December  2020         113       maps    12
79      March  2021         132       maps     3
80      April  2021         148       maps     4
81       June  2021         151       maps     6

[82 rows x 5 columns]

In [118]:
print(maps_df.shape)

(82, 5)


In [119]:
import matplotlib.pyplot as plt
import matplotlib.ticker as mtick
import seaborn as sns
import numpy as np

In [148]:
mapsdf = maps_df.groupby(['month_name'])['query_count'].sum().reset_index()
mapsdf

month_name               query_count
0       April  152130115351189110114148
1      August      11881137238186106125
2    December            97117108121113
3    February         14012913023598100
4     January      97119117104173160154
5        July     121120149236173124139
6        June  121148125265154116136151
7       March       1448333118888124132
8         May     119109125266156105139
9    November        821021201551649686
10    October     100110138224166116127
11  September         11511018122224083

In [147]:
(maps_df.groupby(['month_name'], as_index = False)['query_count']
        .sum()
        .sort_values('query_count', ascending = False)
)

month_name               query_count
4     January      97119117104173160154
2    December            97117108121113
9    November        821021201551649686
0       April  152130115351189110114148
7       March       1448333118888124132
3    February         14012913023598100
6        June  121148125265154116136151
5        July     121120149236173124139
8         May     119109125266156105139
1      August      11881137238186106125
11  September         11511018122224083
10    October     100110138224166116127